src/backend/egl/display.rs

```diff
@@ -0,0 +1,618 @@
+            }
+
+            let mut images = Vec::with_capacity(format.num_planes());
+            for i in 0..format.num_planes() {
+                let mut out = Vec::with_capacity(3);
+                out.push(ffi::egl::WAYLAND_PLANE_WL as i32);
+                out.push(i as i32);
+                out.push(ffi::egl::NONE as i32);
+
+                images.push({
+                    let image = unsafe {
+                        ffi::egl::CreateImageKHR(
+                            *display,
+                            ffi::egl::NO_CONTEXT,
+                            ffi::egl::WAYLAND_BUFFER_WL,
+                            buffer.as_ref().c_ptr() as *mut _,
+                            out.as_ptr(),
+                        )
+                    };
+                    if image == ffi::egl::NO_IMAGE_KHR {
+                        return Err(BufferAccessError::EGLImageCreationFailed);
+                    } else {
+                        image
+                    }
+                });
+            }
+
+            Ok(EGLImages {
+                display: Arc::downgrade(&display),
+                width: width as u32,
+                height: height as u32,
+                y_inverted: inverted != 0,
+                format,
+                images,
+                buffer,
+                #[cfg(feature = "renderer_gl")]
+                gl: self.gl.clone(),
+                #[cfg(feature = "renderer_gl")]
+                egl_to_texture_support: self.egl_to_texture_support,
+            })
+        } else {
+            Err(BufferAccessError::ContextLost)
+        }
+    }
+
+    /// Try to receive the dimensions of a given [`WlBuffer`].
+    ///
+    /// In case the buffer is not managed by EGL (but e.g. the [`wayland::shm` module](::wayland::shm)) or the
+    /// context has been lost, `None` is returned.
+    pub fn egl_buffer_dimensions(&self, buffer: &WlBuffer) -> Option<(i32, i32)> {
+        if let Some(display) = self.display.upgrade() {
+            let mut width: i32 = 0;
+            if unsafe {
+                ffi::egl::QueryWaylandBufferWL(
+                    *display,
+                    buffer.as_ref().c_ptr() as *mut _,
+                    ffi::egl::WIDTH as i32,
+                    &mut width as *mut _,
+                ) == 0
+            } {
+                return None;
+            }
+
+            let mut height: i32 = 0;
+            if unsafe {
+                ffi::egl::QueryWaylandBufferWL(
+                    *display,
+                    buffer.as_ref().c_ptr() as *mut _,
+                    ffi::egl::HEIGHT as i32,
+                    &mut height as *mut _,
+                ) == 0
+            } {
+                return None;
+            }
+
+            Some((width, height))
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(feature = "use_system_lib")]
+impl Drop for WaylandEGLDisplay {
+    fn drop(&mut self) {
+        if let Some(display) = self.display.upgrade() {
```

Comment:
See `EGLContext` drop. We definitely always want to be able to unbind. Otherwise we might end up leaving `wl_drm` in a bad state, which is bad for long running compositors, which might swap their devices, when new graphic systems appear.

Reply:
I'm unsure what would happen if the device that is providing the egl context dies/is removed. If the hardware device disappears, is the `ffi::EGLDisplay` is automatically killed driver side? meaning we'd be unable to call `UnbindWaylandDisplayWL` as the display ptr would be invalid. But, if the display ptr is still valid, then we would be able to call unbind. I'm unsure what the behaviour is around this.

It is also possible that the driver automatically removes the interfaces it has added.

Reply:
This is actually a very interesting question, because I was not yet able to forcefully remove an initialized graphics card from smithay (and I bet it would crash). I guess we would notice this in the udev-handler or through drm-errors (which all should be not critical) very quickly and start cleaning up the device.

But if we are in the middle of a render, GL or EGL might just break and I am pretty confident, that there is an unwrap or a panic somewhere in the rendering code to make smithay crash.

But I guess I cannot reasonably debug this without something like an external thunderbolt gpu.

What I meant instead is if the compositor writer adds a way to remove and EGL- and all underlying devices during runtime (something like an `eject` command for e.g. a thunderbold gpu). The `EGLDisplay` would be destroyed with the `EGLDevice` and the `EGLSurface` would also long be gone, but the `EGLImages` or the `EGLBufferReader` might still be alive until it gets cleaned up in the next rendering step.

And until that `wl_drm` would maybe be in a bad state depending how mesa handles destruction of a bound `EGLDisplay`. Which is why I would also add a strong pointer to those two types to ensure the `EGLDisplay` always dies at last.